[sparsity] Lambda Scheduler

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60728 [quant][sparsity] Generic convert function
* **#59771 [sparsity] Lambda Scheduler**
* #59770 [sparsity] Base sparsity level scheduler class
* #58955 [sparsity] WeightNormSparsifier
* #58707 [sparsity][refactor] Import factoring out
* #58704 [sparsity] Sparsifier class
* #58705 [sparsity] Sparsity parametrization
* #60850 [sparsity][refactor] Changing linear row/col control
* #60887 [sparsity] Add sparsity tests to run_test.py

Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

Commit:
[sparsity] Lambda Scheduler

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"




[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"




[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).

Commit:
Update on "[sparsity] Lambda Scheduler"


Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).

Commit:
Update on "[sparsity] Lambda Scheduler"


Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"


Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).

Commit:
Update on "[sparsity] Lambda Scheduler"


Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"




Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).

Commit:
Update on "[sparsity] Lambda Scheduler"




Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"




Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"




Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"




Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"




Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).

raghuramank100 commented on test/ao/sparsity/test_scheduler.py:
```diff
@@ -38,3 +38,15 @@ def test_step(self):
+        scheduler = LambdaSL(sparsifier, lambda epoch: epoch * 10)
+        assert sparsifier.module_groups[0]['sparsity_level'] == 0.0  # Epoch 0
+        scheduler.step()
+        assert sparsifier.module_groups[0]['sparsity_level'] == 5.0  # Epoch 1
```
Should we ensure that sparsity level is always between 0 and 1 for all schedulers?

z-a-f replied:
Currently the check is:
```
if sparsity_level >= 1.0:
    return torch.zeros(...)
elif sparsity_level <= 0.0:
    return torch.ones(...)
else:
    return weight * mask
```

I can add explicit checks to communicate to the user -- will be part of the base_sparsifier PR

raghuramank100 commented on torch/ao/sparsity/scheduler/lambda_scheduler.py:
```diff
@@ -0,0 +1,39 @@
+from .base_scheduler import BaseScheduler
+
+class LambdaSL(BaseScheduler):
```
Another scheduler that would be very useful to implement is a delayedScheduler. This scheduler will keep sparsity at 0% till a certain number of epochs and then turn it on based on a lambda function, in effect the lambda fn would take two args, epoch, epoch_start and return a sparsity level of 0 if epoch < epoch_start

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).

Commit:
Update on "[sparsity] Lambda Scheduler"


Implements a specific sparsity scheduler, that uses a user-provided lambda's to change the levels.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070604](https://our.internmc.facebook.com/intern/diff/D29070604)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070604).
